File: 131-palindrome-partitioning/131-palindrome-partitioning.cpp
class Solution {
public:
    vector<vector<string>> partition(string s) {
        vector<vector<string>> ans;
        vector<string> part;
        int n = s.length();
        fun(s,ans, part,0);
        return ans;
    }
private:
    void fun(string &s, vector<vector<string>> &ans, vector<string> &part, int i) {
        if (i >= size(s)) {
            ans.push_back(part);
            return;
        }
        // genrate all the possible substrings
        for (int j = i;j < size(s);j++) {
            if (isPalin(s, i, j)) {
                part.push_back(s.substr(i, (j-i +1)));
                fun(s, ans, part, j+1);
                part.pop_back();
            }
        }
    } 
    bool isPalin(string &s, int low, int high) {
        while (low <= high) {
            if (s[low++] != s[high--]) return false;
        }
        return true;
    }
};

File: 131-palindrome-partitioning/131-palindrome-partitioning.py
class Solution:
    def partition(self, s: str):
        tmp, res = [],[]
        def isPalin(i, j):
            while i < j:
                if s[i] != s[j]:
                    return False
                i+=1
                j-=1
            return True
        def helper(start):
            if start >= len(s):
                res.append(tmp[:])
                return
            for i in range(start,len(s)):
                if isPalin(start, i):
                    tmp.append(s[start:i+1])
                    helper(i+1)
                    tmp.pop()
        helper(0)
        return res
                
        
"""
ab

a b
a b

"""